lensless/mask.py

```diff
@@ -0,0 +1,321 @@
+from perlin_numpy import generate_perlin_noise_2d
+from sympy.ntheory import quadratic_residues
+from scipy.signal import max_len_seq
+from waveprop.fresnel import fresnel_conv
+from waveprop.rs import angular_spectrum
+
+
+
+
+
+class Mask(abc.ABC):
+    """
+    Parent Mask class
+    """
+
+    def __init__(self, sensor_size_px: tuple, 
+                 sensor_size_m: tuple, 
+                 feature_size: tuple, 
+                 distance_sensor: float, 
+                 wavelength: float) -> None:
+        """
+        Parent mask contructor.
+        Attributes common to each type of mask.
+
+        Parameters
+        ----------
+        sensor_size_px: tuple (dim=2)
+            size of the sensor (px)
+        sensor_size_m: tuple (dim=2)
+            size of the sensor (m)
+        feature_size: float
+            size of the feature (m)
+        distance_sensor: float
+            distance between the mask and the sensor (m)
+        wavelength: float
+            wavelength (m)
+        """
+        
+        self.mask = None
+        self.psf = None
+        self.phase_mask = None
+        self.height_map = None
+        self.sensor_size_px = sensor_size_px
+        self.sensor_size_m = sensor_size_m
+        self.feature_size = feature_size
+        self.distance_sensor = distance_sensor
+        self.wavelength = wavelength
+        self.create_mask()
+        self.compute_psf()
+    
+
+    @abc.abstractmethod
+    def create_mask(self):
+        """
+        Abstract mask creation method.
+        Creating mask with subclass-specific function.
+        """
+        pass
+
+
+    def compute_psf(self):
+        """
+        Computing the PSF.
+        Common to all types of masks.
+        """
+        self.psf, _, _ = angular_spectrum(
+            u_in=self.mask,
+            wv=self.wavelength,
+            d1=self.feature_size,
+            dz=self.distance_sensor,
+            dtype=np.float32
+        )
+        pass
+    
+
+    def shape(self):
+        """
+        Shape of the mask.
+        """
+        return self.mask.shape
+    
+
+    def phase_retrieval(self, lambd, d1, dz, n=1.5, n_iter=10):
+        """
+        Iterative phase retrieval algorithm from the PhlatCam article (https://ieeexplore.ieee.org/document/9076617)
+
+        Parameters
+        ----------
+        lambd: float
+            wavelength (m)
+        d1: float
+            sample period on the sensor i.e. pixel size (m)
+        dz: float
+            propagation distance between the mask and the sensor
+        n: float
+            refractive index of the mask substrate
+        n_iter: int
+            number of iterations
+        """
+        M_p = np.sqrt(self.psf)
+        for _ in progressbar.ProgressBar()(range(n_iter)):
+            M_phi = np.exp(1j * np.angle(fresnel_conv(M_p, lambd, d1, -dz, dtype=np.float32)[0]))
+            M_p = np.sqrt(self.psf) * np.exp(1j * fresnel_conv(M_phi, lambd, d1, dz, dtype=np.float32)[0])    
```

Comment:
I think you're missing an `np.angle` at the output of Fresnel propagation?

I would recommed breaking up the implementation as in the steps of Algorithm 1 of the paper (and with comments) so that it's more readable.